persist-cli/src/main/java/io/ballerina/persist/nodegenerator/BalSyntaxGenerator.java

```diff
@@ -411,46 +411,92 @@ public static void inferRelationDetails(Module entityModule) {
-                            field.getRelation().setReferences(assocEntity.getKeys());
+                        // Check and avoid having one directional association within entities. Can remove this once
+                        // this is validated at compile time.
+                        long count = assocEntity.getFields().stream().filter(assocfield -> assocfield.getFieldType()
+                                .equals(entity.getEntityName())).count();
+                        if (count == 0) {
+                            throw new RuntimeException(String.format(
+                                    "Bidirectional mapping for the associate field %s is not found in the entity %s",
+                                    field.getFieldName(), assocEntity.getEntityName()));
                         }
+                        if (field.getRelation() == null) {
+                            // this branch only handles one-to-many or many-to-many with no relation annotations
+                            assocEntity.getFields().stream().filter(assocfield -> assocfield.getFieldType()
+                                            .equals(entity.getEntityName()))
+                                    .filter(assocfield -> assocfield.getRelation() == null).forEach(assocfield -> {
+                                        // one-to-many or many-to-many with no relation annotations
+                                        if (!field.isArrayType() && !assocfield.isArrayType()) {
+                                            throw new RuntimeException("Couldn't find the relation owner between " +
+                                                    entity.getEntityName() + " and " + assocEntity.getEntityName());
+                                        }
+                                        field.setRelation(computeRelation(field, entity, assocEntity));
+                                        assocfield.setRelation(computeRelation(assocfield, assocEntity, entity));
+                                    });
+                        } else if (field.getRelation() != null && field.getRelation().isOwner()) {
+                            field.getRelation().setRelationType(field.isArrayType() ?
+                                    Relation.RelationType.MANY : Relation.RelationType.ONE);
+                            field.getRelation().setAssocEntity(assocEntity);
+                            List<String> keyColumns = field.getRelation().getKeyColumns();
+                            if (keyColumns == null || keyColumns.size() == 0) {
+                                keyColumns = assocEntity.getKeys().stream().map(key ->
+                                        assocEntity.getEntityName().toLowerCase(Locale.getDefault())
+                                                + key.substring(0, 1).toUpperCase(Locale.getDefault())
+                                                + key.substring(1)).collect(Collectors.toList());
```

Comment:
Sorry I didn't get the issue here. Can you clarify a bit more

Reply:
```
assocEntity.getEntityName().toLowerCase(Locale.getDefault())
                                                + key.substring(0, 1).toUpperCase(Locale.getDefault())
                                                + key.substring(1))
```
This basically get the first letter capitalised Entity name. This can be a separate property in the entity model. Or else we will capitalise the entity name wherever we need it.

Reply:
got it. I will do the improvement